src/vdaf/mastic.rs

```diff
@@ -379,32 +382,477 @@ where
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    type PrepareState = MasticPrepareState<T::Field, SEED_SIZE>;
+    type PrepareShare = MasticPrepareShare<T::Field, SEED_SIZE>;
+    type PrepareMessage = MasticPrepareMessage<T::Field, SEED_SIZE>;
+
+    fn prepare_init(
+        &self,
+        verify_key: &[u8; SEED_SIZE],
+        agg_id: usize,
+        agg_param: &MasticAggregationParam,
+        nonce: &[u8; NONCE_SIZE],
+        public_share: &MasticPublicShare<VidpfWeight<T::Field>>,
+        input_share: &MasticInputShare<T::Field, SEED_SIZE>,
+    ) -> Result<
+        (
+            MasticPrepareState<T::Field, SEED_SIZE>,
+            MasticPrepareShare<T::Field, SEED_SIZE>,
+        ),
+        VdafError,
+    > {
+        let id = match agg_id {
+            0 => Ok(VidpfServerId::S0),
+            1 => Ok(VidpfServerId::S1),
+            _ => Err(VdafError::Uncategorized(
+                "Invalid aggregator ID".to_string(),
+            )),
+        }?;
+        let mut xof = P::init(
+            verify_key,
+            &self.domain_separation_tag(DST_PATH_CHECK_BATCH),
+        );
+        let mut output_shares = Vec::<T::Field>::with_capacity(
+            self.vidpf.weight_parameter * agg_param.level_and_prefixes.prefixes().len(),
+        );
+        let mut cache_tree = BinaryTree::<VidpfEvalCache<VidpfWeight<T::Field>>>::default();
+        let cache = VidpfEvalCache::<VidpfWeight<T::Field>>::init_from_key(
+            &id,
+            &input_share.vidpf_key,
+            &self.vidpf.weight_parameter,
+        );
+        cache_tree
+            .insert(Path::empty(), cache)
+            .expect("Should alwys be able to insert into empty tree at root");
+        for prefix in agg_param.level_and_prefixes.prefixes() {
+            let mut value_share = self.vidpf.eval_with_cache(
+                &id,
+                &input_share.vidpf_key,
+                public_share,
+                prefix,
+                &mut cache_tree,
+                nonce,
+            )?;
+            xof.update(&value_share.proof);
+            output_shares.append(&mut value_share.share.0);
+        }
+        let root_share_opt = if agg_param.require_weight_check {
+            Some(self.vidpf.eval_root_with_cache(
+                &id,
+                &input_share.vidpf_key,
+                public_share,
+                &mut cache_tree,
+                nonce,
+            )?)
+        } else {
+            None
+        };
+
+        let szk_verify_opt = if let Some(root_share) = root_share_opt {
+            Some(self.szk.query(
+                root_share.as_ref(),
+                input_share.proof_share.clone(),
+                verify_key,
+                nonce,
+            )?)
+        } else {
+            None
+        };
+        let (prep_share, prep_state) =
+            if let Some((szk_query_share, szk_query_state)) = szk_verify_opt {
```

Comment:
Suggestion: It would improve readability if we could remove this `if`-`else` block. To do this we would need to compute `szk_query_state_opt`, `verifier_len`, and `szk_query_share_opt` before we get to this point. I believe we might be able to accomplish this with a couple changes:

1. move `verifier_len` into `szk_query_state` (this may also fix the `match` statements above)
2. replace `let szk_verify_opt = ...` above with `let (szk_query_state_opt, szk_query_share_opt) = ...`